File: Csharp/CodeWars/src/Resolutions/ReverseOrRotate.cs
namespace CodeWars.Resolutions;

public class ReverseOrRotate
{
    public static string RevRot(string s, int sz)
    {
        if (string.IsNullOrWhiteSpace(s) || sz <= 0 || sz > s.Length)
            return string.Empty;

        var output = new List<string>();
        int len = s.Length / sz;

        for (int index = 0; index < len; index++)
        {
            var slice = s.AsSpan(index * sz, sz);


            var arrayValue = new List<int>(sz);
            foreach (var ch in slice)
                arrayValue.Add(ch - '0');

            bool isEven = arrayValue.Sum() % 2 == 0;

            if (isEven)
            {
                arrayValue.Reverse();
                output.Add(string.Concat(arrayValue));
            }
            else
            {
                int first = arrayValue[0];
                arrayValue.RemoveAt(0);
                arrayValue.Add(first);

                output.Add(string.Concat(arrayValue));
            }
        }

        return string.Concat(output);
    }
}

File: TypeScript/Codewars/src/resolutions/Reverse-or-rotate.ts
console.log(revRot("733049910872815764", 5));

export function revRot(s: string, sz: number): string {
	if (s.trim() === "" || sz <= 0 || sz > s.length) return "";

	const output: string[] = [];

	const len = Math.trunc(s.length / sz);

	for (let index = 0; index < len; index++) {
		const arrayValue = s
			.substring(index * sz, index * sz + sz)
			.split("")
			.map(Number);

		const isEven = arrayValue.reduce((prev, curr) => prev + curr, 0) % 2 === 0;

		if (isEven) output.push(arrayValue.reverse().join(""));
		else {
			arrayValue.push(arrayValue.shift() ?? 0);

			output.push(arrayValue.join(""));
		}
	}

	return output.join("");
}
